File: java/20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(char a: s.toCharArray()){
            if(a == '('){
                stack.push(')');
            }else if(a == '{'){
                stack.push('}');
            }else if(a == '['){
                stack.push(']');
            }else{
                // another special case
                if(stack.isEmpty()) 
                    return false;
                    
                if(stack.peek() != a)
                    return false;
                else{
                    stack.pop();
                }
            }

        }
        return stack.isEmpty();

        
    }
}
// @lc code=end

File: js/js/20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
const pair = {
    '(': ')',
    '{': '}',
    '[': ']'
}
var isValid = function (s) {
    const stack = [];
    let i=0
    while (i < s.length) {
        switch (s[i]) {
            case '(':
            case '{':
            case '[':
                stack.push(s[i])
                break;
            case ')':
            case '}':
            case ']':
                if (validate(s[i], stack))
                    stack.pop()
                else
                    return false;
            default:
                return false;
        }
    }
    return i === s.length;
};
function validate(ch, stack) {
     return pair[ch] === stack[stack.length-1]
}
// @lc code=end

